### DISCUSSION TITLE
`authorizationServicesEnabled` reset on client update throug Admin REST API

### DISCUSSION THREAD

**Discussion by lucacavazzana** [keycloak/keycloak#21848] [Q&A]
Hi all,
I have a client configured with the authorization support enabled (`"authorizationServicesEnabled": true`).
**EDIT**: Keycloak 20.0.3 at the time of the post

When updating the configuration using the Admin REST API (adding a `redirectUri` in my case) the authorization is disabled and the settings (what you would see under the _Authorization_ tab of the console) are removed.

```bash
# Headers removed for readability
> curl -XGET "http://keycloak:8080/auth/admin/realms/${REALM}/clients/${CLIENTID}" | jq '.|{authorizationServicesEnabled,serviceAccountsEnabled,redirectUris}'
{
  "authorizationServicesEnabled": true,
  "serviceAccountsEnabled": true,
  "redirectUris": [ ...old... ]
}

> curl -XPUT  "http://keycloak:8080/auth/admin/realms/${REALM}/clients/${CLIENTID}" \
  -d '{"redirectUris":[ ...updated... ]}'

> curl -XGET "http://keycloak:8080/auth/admin/realms/${REALM}/clients/${CLIENTID}" | jq '.|{authorizationServicesEnabled,serviceAccountsEnabled,redirectUris}'
{
  "authorizationServicesEnabled": null,
  "serviceAccountsEnabled": true,
  "redirectUris": [ ...updated... ]
}
```

Trying to restore the value (`PUT`ting `{"authorizationServicesEnabled":true}`) results in Keycloak complaining about not having a service account (despite the `Service account roles` section in the console pointing to a valid SA user):
> ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Uncaught server error: java.lang.RuntimeException: Client does not have a service account.

The only way to reenable the authorization for the client seems to be through the admin console. Still, the settings (resources, policies, permissions,...) seem gone for good and must be re-configured from scratch.

Could this be a bug?

**Comment by gitthonandrhub** [comment]
I recognized this behaviour in keycloak 17 and 21.1.2
It's very bad. On thursday I need to update several customer settings via a script that simply increase the access token life time. A test shows that the complete authorization data is lost after this update.
I'm glad to notice this before I run the update script on the sites.

I know 17 is out of scope, but newer versions should not have this problem.

**Comment by coryanderson-dev** [comment]
Currently hitting this in 24.0.4 with no workaround found

**Reply by gitthonandrhub** [reply]
Thank you for the response.

**Reply by mike-taylor-nomihealth** [reply]
Opened https://github.com/keycloak/keycloak/issues/32530 as a bug from this discussion
